Aztec/Classes/Public/Extensions/NSAttributedString+Lists.swift

```diff
@@ -0,0 +1,152 @@
+    }
+
+
+    /// Return the contents of a TextList following the specified index (inclusive).
+    /// Used to retrieve list items that need to be renumbered.
+    ///
+    /// - Parameter index: An index intersecting a TextList.
+    ///
+    /// - Returns: An NSAttributedString optional containing the list from the specified range or nil if no list was found.
+    ///
+    func textListContents(followingIndex index: Int) -> NSAttributedString? {
+        guard let listRange = rangeOfTextList(atIndex: index) else {
+            return nil
+        }
+
+        let diff = index - listRange.location
+        let subRange = NSRange(location: index, length: listRange.length - diff)
+        return attributedSubstringFromRange(subRange)
+    }
+
+
+    /// Returns the TextList attribute at the specified NSRange, if any.
+    ///
+    /// - Parameter index: The index at which to inspect.
+    ///
+    /// - Returns: A TextList optional.
+    ///
+    func textListAttribute(atIndex index: Int) -> TextList? {
+        return attribute(TextList.attributeName, atIndex: index, effectiveRange: nil) as? TextList
+    }
+
+
+    /// Finds the paragraph ranges in the specified string intersecting the specified range.
+    ///
+    /// - Parameters range: The range within the specified string to find paragraphs.
+    ///
+    /// - Returns: An array containing an NSRange for each paragraph intersected by the specified range.
+    ///
+    func paragraphRanges(spanningRange range: NSRange) -> [NSRange] {
+        var paragraphRanges = [NSRange]()
+        let targetRange = rangeOfEntireString
+
+        let foundationString = string as NSString
+        foundationString.enumerateSubstringsInRange(targetRange, options: .ByParagraphs) { (substring, substringRange, enclosingRange, stop) in
+            // Stop if necessary.
+            if enclosingRange.location >= NSMaxRange(range) {
+                stop.memory = true
+                return
+            }
+
+            // Bail early if the paragraph precedes the start of the selection
+            if NSMaxRange(enclosingRange) <= range.location {
+                return
+            }
+
+            paragraphRanges.append(enclosingRange)
+        }
+
+        return paragraphRanges
+    }
+
+
+    /// Returns all of the paragraphs, spanning at the specified index, with the given TextList Kind.
+    ///
+    /// - Parameters:
+    ///     - index: The index at which to inspect.
+    ///     - kind: The type of TextList.
+    ///
+    /// - Return: A NSRange collection containing the paragraphs with the specified TextList Kind.
+    ///
+    func paragraphRanges(atIndex index: Int, ofListKind kind: TextList.Kind) -> [NSRange] {
+        guard index >= 0 && index < length else {
+            return []
+        }
+
+        guard let range = rangeOfTextList(atIndex: index) else {
+            return []
+        }
+
+        guard let list = textListAttribute(atIndex: index) where list.kind == kind else {
```

Comment:
Since all guards return the same result, maybe you can put all conditions in a single guard?